### PR TITLE
Server-side's calculation function

### PR DESCRIPTION
**Pull request by edumorlom**
Its input is {date->number}, population, data_type.
data_type can be either "Cases" or "Deaths".

and it returns  {
        '7DayAverage' + data_type: week_ave,
        '7DayAverage' + data_type + 'Value': week_ave_value,
        'perCapita' + data_type: per_capita,
        'perCapita' + data_type + 'Value': per_capita_value,
        'pctChange' + data_type + 'Value': pct_increase_value
    }

**To speed up the download, we only return the time_series every even date.
The graph looks the same, but with half the data.**

NOTE: we only return the entire time-series when needed (graphs). Otherwise we only return the value for the date we are observing, aka the latest date.

**7DayAverageValue and perCapitaValue could very well be calculated in the front-end, but I think it looks simpler if we always return a value for average, perCapita and pctChange and time-series is optional.**

### CODE REVIEW

File: covid19-dashboard_NEW/calculate_data.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    pct_increase_value: Union[float, None] = None
+
+    # If there are less elements than our chunk_size.
+    # We can't perform any calculations.
+    if len(data.keys()) > chunk_size:
+        # The latest date is the largest date from the list of dates.
+        latest_date: str = max(list(data.keys()))
+
+        # Calculate the time-series moving average.
+        week_ave: dict = _moving_average(data, chunk=chunk_size)
```
Comment (beets) [inline covid19-dashboard_NEW/calculate_data.py:134]: consider using names that is more different here. week_ave could read as a single value -- maybe week_moving_averages?

File: covid19-dashboard_NEW/calculate_data.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        # Get the value for the current date.
+        week_ave_value: int = week_ave[latest_date]
```
Comment (beets) [inline covid19-dashboard_NEW/calculate_data.py:136]: also consider giving a better name here (this is the latest_weekly_average)?

File: covid19-dashboard_NEW/calculate_data.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            }
+            per_capita_value: float = per_capita[latest_date]
+
+        # Calculate the latest_date - 7 days.
+        date_1_week_ago: datetime = \
+            datetime.fromisoformat(latest_date) - timedelta(days=7)
+        date_1_week_ago: str = date_1_week_ago.strftime('%Y-%m-%d')
+
+        # If the date 1 week ago is present in the dataset.
+        if date_1_week_ago in week_ave:
```
Comment (beets) [inline covid19-dashboard_NEW/calculate_data.py:152]: you need both 1 week previous and latest date to be present.

File: covid19-dashboard_NEW/calculate_data.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        # If population is larger than 0, perform the division on dataset.
+        if population > 0:
+            per_capita: dict = {
+                key: (week_ave[key] / population) * 1_000_000
+                for key in week_ave
+            }
+            per_capita_value: float = per_capita[latest_date]
+
+        # Calculate the latest_date - 7 days.
+        date_1_week_ago: datetime = \
```
Comment (beets) [inline covid19-dashboard_NEW/calculate_data.py:147]: on my first reading this, this was 1 week ago from today, though latest_date could be any date and is simply the last date in the time series.

File: covid19-dashboard_NEW/calculate_data.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+            datetime.fromisoformat(latest_date) - timedelta(days=7)
+        date_1_week_ago: str = date_1_week_ago.strftime('%Y-%m-%d')
+
+        # If the date 1 week ago is present in the dataset.
+        if date_1_week_ago in week_ave:
+            # Get the value for that date.
+            week_ave_1w_value: int = week_ave[date_1_week_ago]
+            # If the value is larger than 0, perform pct change.
+            if week_ave_1w_value > 0:
```
Comment (beets) [inline covid19-dashboard_NEW/calculate_data.py:156]: can we better handle these negative values? if they are corrective, can we perform the correction and ensure positive numbers throughout?

File: covid19-dashboard_NEW/calculate_data.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    :param population: geoId -> population.
+    :return: a dictionary with the 7-week moving average + other calculations.
+    """
+    # For example, 1-week moving average.
+    chunk_size: int = 7
+
+    # Instantiate the variables to {} or None depending on data type.
+    week_ave: dict = {}
+    week_ave_value: Union[float, None] = None
+    per_capita: dict = {}
```
Comment (beets) [inline covid19-dashboard_NEW/calculate_data.py:123]: what does a per_capita dictionary mean?

File: covid19-dashboard_NEW/calculate_data.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    per_capita_value: Union[float, None] = None
+    pct_increase_value: Union[float, None] = None
+
+    # If there are less elements than our chunk_size.
+    # We can't perform any calculations.
+    if len(data.keys()) > chunk_size:
+        # The latest date is the largest date from the list of dates.
+        latest_date: str = max(list(data.keys()))
```
Comment (beets) [inline covid19-dashboard_NEW/calculate_data.py:131]: what if latest_date is quite far in the past for this geo?

File: covid19-dashboard_NEW/calculate_data.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+            week_ave_1w_value: int = week_ave[date_1_week_ago]
+            # If the value is larger than 0, perform pct change.
+            if week_ave_1w_value > 0:
+                pct_increase_value: float = \
+                    ((week_ave_value / week_ave_1w_value) - 1) * 100
+
+        # Only keep even indices.
+        # This step is done to improve file size.
+        week_ave = _clean_dataset(week_ave, 2)
+        per_capita = _clean_dataset(week_ave, 2)
```
Comment (beets) [inline covid19-dashboard_NEW/calculate_data.py:163]: should be _clean_dataset(per_capita, 2)

File: covid19-dashboard_NEW/calculate_data.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+
+
+def calculate_data(data_type: str, data: dict, population: int) -> dict:
+    """
+    Given a dataset for a region and its population,
+    perform different types of calculations on every date.
+    :param data_type: type of data, "cases" or "deaths".
+    :param data: a dictionary containing date->int.
+    :param population: geoId -> population.
+    :return: a dictionary with the 7-week moving average + other calculations.
```
Comment (beets) [inline covid19-dashboard_NEW/calculate_data.py:115]: please be explicit about the structure of the return dict.

**Comment by beets** [comment]
Any updates here, or should we close this PR? The comments still stand though.

**Comment by edumorlom** [comment]
Still working on this PR. Will have an update soon. Don’t close it. Thanks!

File: covid19-dashboard_NEW/calculate_data.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+    # The latest date is the largest date from the list of dates.
+    latest_date: str = max(list(date_to_val.keys()))
+
+    # Get a list of (Key, Value).
+    # Key = date, Value = cases/deaths.
+    date_to_val_list = list(date_to_val.items())
+    # Sort the values by date, elem[0].
+    date_to_val_list = sorted(date_to_val_list, key=lambda e: e[0])
+
+    # Only include the element if its index is divisible by chunk.
```
[the remark this answers is not in the document]
Comment (edumorlom) [reply covid19-dashboard_NEW/calculate_data.py:80]: Fixed.

File: covid19-dashboard_NEW/calculate_data.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+    date_to_val_list = date_to_val_list[::step_size]
+
+    # Remove dates before March 1st.
```
[the remark this answers is not in the document]
Comment (edumorlom) [reply covid19-dashboard_NEW/calculate_data.py:83]: Fixed.

File: covid19-dashboard_NEW/calculate_data.py (hunk 2 above)
Comment (edumorlom) [reply covid19-dashboard_NEW/calculate_data.py:136]: Fixed.

File: covid19-dashboard_NEW/calculate_data.py (hunk 8 above)
Comment (edumorlom) [reply covid19-dashboard_NEW/calculate_data.py:163]: Fixed.

File: covid19-dashboard_NEW/calculate_data.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+    :param chunk: number of days to take the average of. 1 week is the default.
+    :return: a new dictionary, date->int where date is the average
+    of the previous chunk days.
+    """
+    # If there are less elements than our chunk_size, then exit.
+    # We can't perform any calculations.
+    if len(date_to_val.keys()) <= chunk:
+        return {}
+
+    # Get a list of [Key, Value].
```
[the remark this answers is not in the document]
Comment (edumorlom) [reply covid19-dashboard_NEW/calculate_data.py:33]: Fixed.

File: covid19-dashboard_NEW/calculate_data.py (hunk 9 above)
Comment (edumorlom) [reply covid19-dashboard_NEW/calculate_data.py:115]: Fixed. Added return type, README.md info and more comments.

File: covid19-dashboard_NEW/calculate_data.py [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+    ]
+
+    # Because the latest date might have been skipped by the step_size.
+    # Check to see if it's in our dataset, otherwise append it.
+    # We need the latest date to be present.
+    if date_to_val_list[-1][0] != latest_date:
+        date_to_val_list.append((latest_date, date_to_val[latest_date]))
+
+    # Convert the list of (date, value) back to a dictionary.
+    output = dict(date_to_val_list)
```
[the remark this answers is not in the document]
Comment (edumorlom) [reply covid19-dashboard_NEW/calculate_data.py:103]: _clean_dataset() and _moving_average() now take DateToValueListType as input.

DateToValueListType = list of tuples of type (date, value).


calculate_data() takes in a dictionary and returns a dictionary.

File: covid19-dashboard_NEW/calculate_data.py [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+
+from datetime import datetime, timedelta
+
+from typing import Union
+
+
+def _moving_average(date_to_val: dict, chunk: int = 7) -> dict:
+    """
+    Method allows us to calculate the time-series average on a dictionary.
+    :param date_to_val: a dictionary where date->int.
```
[the remark this answers is not in the document]
Comment (edumorlom) [reply covid19-dashboard_NEW/calculate_data.py:23]: Fixed.

File: covid19-dashboard_NEW/calculate_data.py [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from datetime import datetime, timedelta
+
+from typing import Union
+
+
+def _moving_average(date_to_val: dict, chunk: int = 7) -> dict:
```
[the remark this answers is not in the document]
Comment (edumorlom) [reply covid19-dashboard_NEW/calculate_data.py:20]: Added.

File: covid19-dashboard_NEW/calculate_data.py [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+        # The latest date is the largest date from the list of dates.
+        latest_date: str = max(list(data.keys()))
+
+        # Calculate the time-series moving average.
+        week_ave: dict = _moving_average(data, chunk=chunk_size)
+        # Get the value for the current date.
+        week_ave_value: int = week_ave[latest_date]
+
+        # If population is larger than 0, perform the division on dataset.
+        if population > 0:
```
[the remark this answers is not in the document]
Comment (edumorlom) [reply covid19-dashboard_NEW/calculate_data.py:139]: If population <= 0, calculate_data now returns {}.

If calculate_data() returns {}, the server will drop the place.

I have added the changes to the main.py

File: covid19-dashboard_NEW/calculate_data.py (hunk 6 above)
Comment (edumorlom) [reply covid19-dashboard_NEW/calculate_data.py:123]: Fixed.

File: covid19-dashboard_NEW/calculate_data.py (hunk 7 above)
Comment (edumorlom) [reply covid19-dashboard_NEW/calculate_data.py:131]: I no longer calculate latest_date in the back-end -- I'm leaving it up to the front-end to pick a date from the time-series. So latest-date is no longer needed in this case.

File: covid19-dashboard_NEW/calculate_data.py (hunk 5 above)
Comment (edumorlom) [reply covid19-dashboard_NEW/calculate_data.py:156]: For example, in this time-series:

Day 1: 10 new deaths
Day 2: -5 new deaths
Day 3: 15 new deaths

-5 usually means they corrected the dataset.
Should we replace Day 2 value with a 0, instead?

File: covid19-dashboard_NEW/calculate_data.py (hunk 1 above)
Comment (edumorlom) [reply covid19-dashboard_NEW/calculate_data.py:134]: Fixed.

File: covid19-dashboard_NEW/calculate_data.py (hunk 4 above)
Comment (edumorlom) [reply covid19-dashboard_NEW/calculate_data.py:147]: I changed to date and baseline_date.

baseline_date = original date
date = current date

Let me know if this still doesn't work.

File: covid19-dashboard_NEW/calculate_data.py (hunk 3 above)
Comment (edumorlom) [reply covid19-dashboard_NEW/calculate_data.py:152]: Algorithm is now different.
Calculates pct_change on the entire timeseries, similar to calculations.
It has its own function now.

**Comment by edumorlom** [comment]
There are three individual commits:
1. Move calculation_data from _NEW to its original folder.
2. Changes + New Unit Tests for moving_average and pct_increase.
3. Documentation about calculations performed in README.md.


Also, I no longer calculate the single values for the latest date and let the front-end pick whatever date it wants.
Pct_increase is now calculated for the entire time_series.

**Comment by beets** [comment]
Please avoid force-pushing if possible so we can keep commit history (and I can only review changes since last commit). Otherwise it's a full review every time.

**Pull request merged**